### PR TITLE
build: refactor Bazel configuration towards multiple runtime tests

### PR DESCRIPTION
**Pull request by mathetake**
- Towards tests on #134, #49, and #113 
- delete variablez.bzl in favor of .bazelrc as we do in Envoy.

### CODE REVIEW

**Comment by mathetake** [comment]
will add WAVM and V8 (and WAMR) builds in following PRs.

File: .bazelrc [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+build:linux --cxxopt=-std=c++17
+build:macos --cxxopt=-std=c++17
+# TODO(mathetake): Windows build is not verified yet.
+# build:windows --cxxopt="/std:c++17"
+
+# LINKOPTS
+# See https://bytecodealliance.github.io/wasmtime/c-api/
+build:linux --action_env=BAZEL_LINKOPTS=-lm:-lpthread:-ldl
+# TODO(mathetake): Windows build is not verified yet.
+# build:windows --action_env=BAZEL_LINKOPTS=ws2_32.lib:advapi32.lib:userenv.lib:ntdll.lib:shell32.lib:ole32.lib
```
Comment (PiotrSikora) [inline .bazelrc:13]: Please group those by config name, not a feature (i.e. group all `build:linux` together, etc.).

File: BUILD [hunk 2]
```diff
@@ -13,27 +18,72 @@ cc_library(
...
+    name = "lib",
+    deps = [
+        ":common_lib",
+    ] + proxy_wasm_select_wasm_runtime_wasmtime(
+        ["wasmtime_lib"],
+    ) + proxy_wasm_select_wasm_runtime_v8(
+        ["v8_lib"],
+    ) + proxy_wasm_select_wasm_runtime_wavm(
+        ["wavm_lib"],
+    ),
```
Comment (PiotrSikora) [inline BUILD:88]: Nit: please sort this alphabetically.

File: bazel/BUILD [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+config_setting(
+    name = "wasm_runtime_v8",
+    values = {"define": "wasm_runtime=v8"},
+)
+
+config_setting(
+    name = "wasm_runtime_wasmtime",
+    values = {"define": "wasm_runtime=wasmtime"},
+)
```
Comment (PiotrSikora) [inline bazel/BUILD:14]: Nit: please sort this alphabetically.

File: bazel/BUILD [hunk 4]
```diff
@@ -0,0 +1,14 @@
+config_setting(
+    name = "wasm_runtime_wavm",
+    values = {"define": "wasm_runtime=wavm"},
```
Comment (PiotrSikora) [inline bazel/BUILD:3]: I think `runtime=XXX` should be self-explanatory in this context, adding `wasm` everywhere adds more noise than value.

File: bazel/select.bzl [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    return select({
+        "@proxy_wasm_cpp_host//bazel:wasm_runtime_wavm": xs,
+        "//conditions:default": [],
+    })
+
+def proxy_wasm_select_wasm_runtime_wasmtime(xs):
+    return select({
+        "@proxy_wasm_cpp_host//bazel:wasm_runtime_wasmtime": xs,
+        "//conditions:default": [],
+    })
```
Comment (PiotrSikora) [inline bazel/select.bzl:31]: Nit: please sort this alphabetically.

File: .github/workflows/cpp.yml [hunk 6]
```diff
@@ -56,17 +56,20 @@ jobs:
 
   build:
     runs-on: ubuntu-latest
-    
+    strategy:
+      matrix:
+        # TODO(mathetake): Add other runtimes.
+        wasm_runtime: [ "wasmtime" ]
```
Comment (PiotrSikora) [inline .github/workflows/cpp.yml:62]: `runtime=XXX` should be self-explanatory in this context, adding `wasm` everywhere adds more noise than value.

File: .github/workflows/cpp.yml [hunk 7]
```diff
@@ -56,17 +56,20 @@ jobs:
 
   build:
     runs-on: ubuntu-latest
-
```
Comment (PiotrSikora) [inline .github/workflows/cpp.yml:59]: Nit: please leave empty lines before and after `strategy`.

**Review by PiotrSikora — COMMENTED** [review]
Some style changes, otherwise LGTM, thanks!

Let's merge WAMR first, since I think it's easier for us to add WAMR here, than to ask WAMR folks to catch-up with our changes again.

File: BUILD [hunk 8]
```diff
@@ -13,27 +18,72 @@ cc_library(
...
+cc_library(
+    name = "v8_lib",
+    srcs = glob([
+        # TODO(@mathetake): Add V8 lib.
+        # "src/v8/*.h",
+        # "src/v8/*.cc",
+    ]),
+    deps = [
+        ":common_lib",
+        # TODO(@mathetake): Add V8 lib.
```
Comment (PiotrSikora) [inline BUILD:49]: FWIW, it's probably not worth spending time too much time on getting V8 to build, and we can wait until [Bazel is officially supported in V8](https://bugs.chromium.org/p/v8/issues/detail?id=11234).

File: .bazelrc [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+# linux
+build:linux --cxxopt=-std=c++17
+# See https://bytecodealliance.github.io/wasmtime/c-api/
+build:linux --action_env=BAZEL_LINKOPTS=-lm:-lpthread:-ldl
+
+# macos
+build:macos --cxxopt=-std=c++17
+
+# windows
```
Comment (PiotrSikora) [inline .bazelrc:11]: Nit: those `linux`, `macos` and `windows` comments are redundant.

File: .bazelrc [hunk 10]
```diff
@@ -0,0 +1,15 @@
+build --enable_platform_specific_config
+
+# linux
+build:linux --cxxopt=-std=c++17
+# See https://bytecodealliance.github.io/wasmtime/c-api/
+build:linux --action_env=BAZEL_LINKOPTS=-lm:-lpthread:-ldl
```
Comment (PiotrSikora) [inline .bazelrc:6]: Would `--linkopt` work here as well, or do we really need `BAZEL_LINKOPTS`?

File: BUILD [hunk 11]
```diff
@@ -13,28 +19,87 @@ cc_library(
...
+cc_library(
+    name = "wamr_lib",
+    srcs = glob([
+        # TODO(@mathetake): Add WAMR lib.
+        # "src/wamr/*.h",
+        # "src/wamr/*.cc",
+    ]),
+    deps = [
+        ":common_lib",
+        # TODO(@mathetake): Add WAMR lib.
```
Comment (PiotrSikora) [inline BUILD:63]: Any chance you could try to add WAMR here? Or do you prefer to leave it for a separate PR?

File: BUILD [hunk 12]
```diff
@@ -13,28 +19,87 @@ cc_library(
...
+cc_library(
+    name = "wavm_lib",
+    srcs = glob([
+        # TODO(@mathetake): Add WAVM lib.
+        # "src/wavm/*.h",
+        # "src/wavm/*.cc",
+    ]),
+    deps = [
+        ":common_lib",
+        # TODO(@mathetake): Add WAVM lib.
```
Comment (PiotrSikora) [inline BUILD:88]: Any chance you could try to add WAVM here? Or do you prefer to leave it for a separate PR?

File: .bazelrc (hunk 10 above)
Comment (mathetake) [reply .bazelrc]: yeah linkopt actually works!

File: BUILD (hunk 11 above)
Comment (mathetake) [reply BUILD:63]: I will do in a separate PR.

File: BUILD (hunk 12 above)
Comment (mathetake) [reply BUILD:88]: I will do in a separate PR.

File: .bazelrc [hunk 13]
```diff
@@ -0,0 +1,12 @@
+build --enable_platform_specific_config
+
+build:linux --cxxopt=-std=c++17
+# See https://bytecodealliance.github.io/wasmtime/c-api/
+build:linux --linkopt="-lm -lpthread -ldl"
```
Comment (PiotrSikora) [inline .bazelrc:5]: Those are separate options, so it should be:
```
build:linux --linkopt=-lm
build:linux --linkopt=-lpthread
build:linux --linkopt=-ldl
```

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Review by PiotrSikora — CHANGES_REQUESTED** [review]
/wait

File: .bazelrc [hunk 14]
```diff
@@ -0,0 +1,14 @@
+build --enable_platform_specific_config
+
+build:linux --cxxopt=-std=c++17
+# See https://bytecodealliance.github.io/wasmtime/c-api/
+build:linux --linkopt=-lm
+build:linux --linkopt=-lpthread
+build:linux --linkopt=-ldl
```
Comment (PiotrSikora) [inline .bazelrc:7]: This `.bazelrc` only applies to builds in this workspace.

Shouldn't those `linkopts` be a part of `wasmtime_lib`, so that other applications that link against it could pick up them up? But even then, I think the embedding binary would need to specify them as well...

Comment (PiotrSikora) [reply .bazelrc:7]: Actually, let's deal with this once it becomes a problem.

**Review by PiotrSikora — APPROVED** [review]
Could you verify that this PR doesn't accidentally break Envoy build before merging? Thanks!

File: .bazelrc (hunk 14 above)
Comment (mathetake) [reply .bazelrc:7]: yeah, thanks for pointing this out. I will keep that in mind.

**Comment by mathetake** [comment]
just verified that Envoy can be built normally against this branch. Now merging. Thanks!
